### ISSUE TITLE
Proposal: Replace 'maxNewSynapseCount' with 'subsampleSize'

### ISSUE THREAD

**Issue by mrcslws** [numenta/htmresearch#671]
The 'newSynapseCount' parameter goes all the way back to TP.py, maybe earlier. In TemporalMemory, it became 'maxNewSynapseCount'.

Maybe its name was correct with earlier versions of the algorithm, but at some point it became something else. It is now essentially the "desired number of active synapses on a segment", or the "subsample size". We've discussed this here: https://github.com/numenta/nupic/issues/3268

In the ColumnPooler, we use this parameter for both proximal and distal synapses. For distal synapses, it behaves as it does in the TemporalMemory -- as a "subsample size". For proximal synapses, it behaves as the name suggests -- as a "max new synapse count".

In my recent change https://github.com/numenta/nupic.research/pull/666, I made it behave as a "max new synapse count" in both places. But @ywcui1990 pointed out that the previous behavior gave us an important capability.

**Proposal: Let's get rid of 'maxNewSynapseCount', and add a 'subsampleSize' parameter**

In other words, let's do what the TemporalMemory does, but give it a better name.

During learning, for proximal and distal connections, it will try to grow `max(0, subsampleSize - numActiveSynapses)` synapses to active input bits. 

After this change, the ColumnPooler will no longer offer a "max new synapse count" feature for distal or proximal synapses. It will always use the 'subsampleSize' to determine the number of synapses to grow.

So we'll have a 'subsampleSizeProximal' and 'subsampleSizeDistal'.

Thoughts, @subutai and @ywcui1990?

Regarding naming: does it make sense to say "subsample" rather than "sample"? We're just sampling the active bits, I'm not sure if that's technically "subsampling". Another option: 'patternSampleSize'

**Comment by subutai** [comment]
👍 for changing the name. I think sampleSize is better (we are sampling and it's a shorter name).  I assume we are always sampling new synapses, even when learning on existing segments?

**Comment by mrcslws** [comment]
When learning on an existing segment, we will only grow synapses if the segment has fewer than 'sampleSize' active synapses. Segments want to have 'sampleSize' active synapses, and they'll grow synapses to make up the difference. They'll sample the active bits that they haven't already synapsed.

**Comment by mrcslws** [comment]
Finally fixed by https://github.com/numenta/htmresearch/pull/761
